src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackDecoder.cs

```diff
@@ -2,15 +2,13 @@
 // Licensed under the Apache License, Version 2.0.
 // See THIRD-PARTY-NOTICES.TXT in the project root for license information.
 
+using System.Buffers;
 using System.Diagnostics;
 
 namespace System.Net.Http.HPack
 {
     internal class HPackDecoder
     {
-        // Can't use Action<,> because of Span
-        public delegate void HeaderCallback(object state, ReadOnlySpan<byte> headerName, ReadOnlySpan<byte> headerValue);
```

Comment:
Delegate dispatch is generally faster than interface dispatch, especially when call sites end up being polymorphic to a number of interface implementations.

Reply:
Does this still hold true? I ran a benchmark yesterday that indicated interfaces (at least, the way we use them here) were faster.

Reply:
Can you point me to the benchmarks? I'm going to have to add some with this PR.

Reply:
Yeah, that surprises me. Please share the benchmark you used. Thanks.

Reply:
Please see https://github.com/scalablecory/delegate-interface/blob/master/delegate-interface/Program.cs

|         Method |     Mean |     Error |    StdDev |
|--------------- |---------:|----------:|----------:|
|  BenchDelegate | 47.27 us | 0.6498 us | 0.6079 us |
| BenchInterface | 20.79 us | 0.3939 us | 0.3685 us |

Reply:
Can we actually do this with the HPackDecoder? We have microbenchmarks in aspnetcore. Might as well add benchmarks here and add a test? https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/perf/Kestrel.Performance/IntegerDecoderBenchmark.cs

Reply:
That benchmark is not comparing apples-to-apples.  The delegate version is doing more work than the interface version.  Change it to:
https://gist.github.com/stephentoub/a75ac6f7337a18df68bd23f4f682d926

Reply:
> That benchmark is not comparing apples-to-apples.

It is doing what `HttpStream` and `HPackDecoder` are doing. If we want to change `HttpStream` to allocate delegates instead of using a static thunk like it currently does, that's fine too.

Reply:
Using an instance delegate, the timing is identical:

|         Method |     Mean |     Error |    StdDev |
|--------------- |---------:|----------:|----------:|
|  BenchDelegate | 20.90 us | 0.2002 us | 0.1873 us |
| BenchInterface | 20.94 us | 0.3214 us | 0.3006 us |

So, it may be worth avoiding the allocation.

Reply:
@jkotalik I agree we should add some general benchmarks. The HPACK code has a lot of room for improvement right now and we need a way to track that. Lets find out where HPACK will live first and work on building it out there.

Reply:
> Using an instance delegate, the timing is identical

On my machine delegates are 15% faster.

Reply:
> On my machine delegates are 15% faster.

That's super interesting. I'll need to take a look at code gen to see what the difference is. Extra level of vtable indirection, maybe.

I suspect such a small change is so minute it will be eaten by all the other code being ran, so I'm comfortable leaving this as delegates or as interfaces for now -- you guys decide :). I think we need a holistic benchmarking/optimization effort in the 5.0 time frame and a data-informed decision could be made at that point.

Reply:
If it doesn't regress throughput, I'm fine with it changing.  I simply wanted to make sure it had been measured and demonstrated not to regress.

Reply:
At what point would you consider something like `HPackDecoder<THttpHeadersHandler> where THttpHeadersHandler : IHttpHeadersHandler` to optimize the dispatch further?

I know this pattern only helps when the interface implementation is a struct, but in the case of the Kestrel HTTP/1.x parser, @benaadams simply wrapped the class (Frame) with a struct (FrameAdapter) which made the indirection explicit. Given that the struct still needs to follow a pointer to the class instance, I wouldn't have expected a perf improvement, but it helped in the case of the Kestrel parser: https://github.com/aspnet/KestrelHttpServer/pull/1640

Reply:
> At what point would you consider something like HPackDecoder<THttpHeadersHandler> where THttpHeadersHandler : IHttpHeadersHandler to optimize the dispatch further?

When it's proven to make a meaningful and impactful improvement to throughput that's worth the added complexity.

Reply:
The struct wrapper turns the interface dispatch via generic constraint into direct calls; turning the slowest type of call (interface+generic) into the fastest type of call (direct).

Reply:
And adds complexity to the type and declaration sites as well as added generic overheads.  Hence the "meaningful" part of my statement.

Reply:
https://gist.github.com/benaadams/a7782d338a67abb71dca0399d76a87ce

|                               Method |      Mean |
|------------------------------------- |----------:|
|                        BenchDelegate | 18.687 us |
|                       BenchInterface | 25.748 us |
| BenchGenericConstraintConcreteStruct |  3.118 us |
```

Reply:
Being more fair and adding NoInlining 
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public int Add(int x, int y) => x + y;
```
as the call will be too big to inline the calling types are

|                               Method |     Mean |     Error |
|------------------------------------- |---------:|----------:|
|                        BenchDelegate | 17.15 us | 0.1082 us |
|                       BenchInterface | 23.08 us | 0.1620 us |
|               BenchGenericConstraint | 23.35 us | 0.3040 us |
|      BenchGenericConstraintInterface | 23.38 us | 0.2970 us |
| BenchGenericConstraintConcreteStruct | 15.16 us | 0.3024 us |

Reply:
Meaningful in context, not in a microbenchmark measuring just the cost of the call.

Reply:
So the reason I changed it to an interface rather than a delegate is that in Kestrel, we use an interface to handle headers for both HTTP/1 and HTTP/2: https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Core/src/Internal/Http/IHttpHeadersHandler.cs
https://github.com/aspnet/AspNetCore/blob/eebd61642f5d3afc6427bb4526b9f248daafd00a/src/Servers/Kestrel/Core/src/Internal/Http/Http1ParsingHandler.cs#L8
https://github.com/aspnet/AspNetCore/blob/2de6c732d1f410460a627cb75aca91ae2c741fd5/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L29

From @benaadams numbers, it does look like interfaces could regress performance. Before this PR is merged, we can either make the optimization to `HPackDecoder<THttpHeadersHandler> where THttpHeadersHandler : IHttpHeadersHandler` or we can make it a follow up work item.

Reply:
My benchmark indicates that versus what we're currently doing (using a delegate to a static method and thunking to an instance method), this new implementation will be faster -- so this change should improve throughput on current state, even if could be further optimized (in a later PR) by using a delegate to instance method or a struct.

Unless I'm missing something, seems like this code is OK as it accomplishes the goal of the PR and doesn't hurt throughput.